src/ChainRulesTestUtils.jl

```diff
@@ -0,0 +1,186 @@
+
+# Arguments
+- `f`: Function for which the `frule` should be tested.
+- `x`: input at which to evaluate `f` (should generally be set to an arbitary point in the domain).
+- `ẋ`: differential w.r.t. `x` (should generally be set randomly).
+
+All keyword arguments except for `fdm` are passed to `isapprox`.
+"""
+function frule_test(f, (x, ẋ); rtol=1e-9, atol=1e-9, fdm=_fdm, kwargs...)
+    return frule_test(f, ((x, ẋ),); rtol=rtol, atol=atol, fdm=fdm, kwargs...)
+end
+
+function frule_test(f, xẋs::Tuple{Any, Any}...; rtol=1e-9, atol=1e-9, fdm=_fdm, kwargs...)
+    ensure_not_running_on_functor(f, "frule_test")
+    xs, ẋs = collect(zip(xẋs...))
+    Ω, dΩ_ad = frule(f, xs..., NO_FIELDS, ẋs...)
+    @test f(xs...) == Ω
+
+    # Correctness testing via finite differencing.
+    dΩ_fd = jvp(fdm, xs->f(xs...), (xs, ẋs))
+    @test isapprox(
+        collect(extern.(dΩ_ad)),  # Use collect so can use vector equality
+        collect(dΩ_fd);
+        rtol=rtol,
+        atol=atol,
+        kwargs...
+    )
+end
+
+
+"""
+    rrule_test(f, ȳ, (x, x̄)...; rtol=1e-9, atol=1e-9, fdm=central_fdm(5, 1), kwargs...)
+
+# Arguments
+- `f`: Function to which rule should be applied.
+- `ȳ`: adjoint w.r.t. output of `f` (should generally be set randomly).
+  Should be same structure as `f(x)` (so if multiple returns should be a tuple)
+- `x`: input at which to evaluate `f` (should generally be set to an arbitary point in the domain).
+- `x̄`: currently accumulated adjoint (should generally be set randomly).
+
+All keyword arguments except for `fdm` are passed to `isapprox`.
+"""
+function rrule_test(f, ȳ, (x, x̄)::Tuple{Any, Any}; rtol=1e-9, atol=1e-9, fdm=_fdm, kwargs...)
+    ensure_not_running_on_functor(f, "rrule_test")
+
+    # Check correctness of evaluation.
+    fx, pullback = rrule(f, x)
+    @test collect(fx) ≈ collect(f(x))  # use collect so can do vector equality
+    (∂self, x̄_ad) = if fx isa Tuple
+        # If the function returned multiple values,
+        # then it must have multiple seeds for propagating backwards
+        pullback(ȳ...)
+    else
+        pullback(ȳ)
+    end
+
+    @test ∂self === NO_FIELDS  # No internal fields
+    # Correctness testing via finite differencing.
+    x̄_fd = j′vp(fdm, f, ȳ, x)
+    @test isapprox(x̄_ad, x̄_fd; rtol=rtol, atol=atol, kwargs...)
+end
+
+# case where `f` takes multiple arguments
+function rrule_test(f, ȳ, xx̄s::Tuple{Any, Any}...; rtol=1e-9, atol=1e-9, fdm=_fdm, kwargs...)
+    ensure_not_running_on_functor(f, "rrule_test")
+
+    # Check correctness of evaluation.
+    xs, x̄s = collect(zip(xx̄s...))
+    y, pullback = rrule(f, xs...)
+    @test f(xs...) == y
+
+    @assert !(isa(ȳ, Thunk))
+    ∂s = pullback(ȳ)
+    ∂self = ∂s[1]
+    x̄s_ad = ∂s[2:end]
+    @test ∂self === NO_FIELDS
+
+    # Correctness testing via finite differencing.
+    x̄s_fd = _make_fdm_call(fdm, f, ȳ, xs, x̄s .== nothing)
```

Comment:
```suggestion
    x̄s_fd = j′vp(fdm, f, ȳ, xs...)
    map(x̄s_ad, x̄s_fd) do x̄_ad, x̄_fd
        @test isapprox(x̄_ad, x̄_fd; rtol=rtol, atol=atol, kwargs...)
    end
```
and remove the `for` loop below this.